File: Sources/practice-harness/FirebaseAuth.swift
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Minimal Firebase Identity Platform REST client for the harness's
/// self-contained sign-in. Ported from the e2e suite's `firebaseAuth.ts` so the
/// runner can authenticate the pinned test user without Node or any prior
/// e2e state.
///
/// Strategy: try the cached refresh token first (cheap,
/// avoids the rate-limited `mfaSignIn:finalize` quota); on ANY failure — 401,
/// expired/revoked refresh token, network error — fall back to the full TOTP
/// MFA dance. Either path returns a fresh, MFA-stamped ID token plus a NEW
/// refresh token (the old one is invalidated within seconds of exchange).
struct FirebaseAuth {
    let apiKey: String

    private let idpBase = "https://identitytoolkit.googleapis.com"
    private let secureTokenBase = "https://securetoken.googleapis.com"

    struct MintResult {
        let idToken: String
        let refreshToken: String
        let mode: String
    }

    enum AuthError: LocalizedError {
        case http(path: String, status: Int, body: String)
        case shape(String)

        var errorDescription: String? {
            switch self {
            case let .http(path, status, body): "\(path) failed: \(status) \(body)"
            case let .shape(message): message
            }
        }
    }

    func mint(
        refreshToken: String?, email: String?, password: String?, totpSecret: String?
    ) async throws -> MintResult {
        if let refreshToken, !refreshToken.isEmpty {
            do {
                let (id, rt) = try await exchangeRefreshToken(refreshToken)
                return MintResult(idToken: id, refreshToken: rt, mode: "refresh-exchange")
            } catch {
                FileHandle.standardError.write(Data(
                    "refresh exchange failed, falling back to TOTP MFA: \(error.localizedDescription)\n".utf8
                ))
            }
        }

        guard let email, let password, let totpSecret,
              !email.isEmpty, !password.isEmpty, !totpSecret.isEmpty
        else {
            throw AuthError.shape("TOTP fallback requires FB_EMAIL, FB_PASSWORD, FB_TOTP_SECRET")
        }
        // The Firebase Auth blocking function is scaled to zero on non-prod
        // projects (cost), so the first mfaSignIn:finalize after it idles down
        // cold-starts and Identity Platform reports "BLOCKING_FUNCTION_ERROR_
        // RESPONSE: Cloud function deadline exceeded". It's warm on the next
        // try. Retry the whole MFA sign-in (fresh pending credential + fresh
        // TOTP each attempt) rather than fail the run on a cold start.
        var lastError: Error?
        for attempt in 1 ... 4 {
            do {
                let (id, rt) = try await signInWithMfa(
                    email: email, password: password, totpSecret: totpSecret
                )
                return MintResult(idToken: id, refreshToken: rt, mode: "totp-mfa")
            } catch let error where Self.isBlockingFunctionColdStart(error) {
                lastError = error
                FileHandle.standardError.write(Data(
                    "mfaSignIn:finalize hit a cold blocking function (attempt \(attempt)/4) — warming, retrying\n".utf8
                ))
                try? await Task.sleep(nanoseconds: 6_000_000_000)
            }
        }
        throw lastError ?? AuthError.shape("MFA sign-in failed after cold-start retries")
    }

    /// True for the transient "blocking function scaled to zero" cold-start
    /// failure — retryable once the function warms. Any other auth error
    /// (bad creds, wrong TOTP, 401) is NOT retried here.
    private static func isBlockingFunctionColdStart(_ error: Error) -> Bool {
        guard case let AuthError.http(_, _, body) = error else { return false }
        return body.contains("BLOCKING_FUNCTION_ERROR_RESPONSE") || body.contains("deadline exceeded")
    }

    // MARK: - Refresh-token exchange (securetoken)

    private func exchangeRefreshToken(_ refreshToken: String) async throws -> (String, String) {
        guard let url = URL(string: "\(secureTokenBase)/v1/token?key=\(apiKey)") else {
            throw AuthError.shape("bad securetoken URL")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data("grant_type=refresh_token&refresh_token=\(refreshToken)".utf8)

        let (data, response) = try await URLSession.shared.data(for: request)
        try ensureOK(response, data, path: "securetoken/v1/token")

        // The endpoint returns snake_case despite the rest of IDP being camelCase.
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        guard let id = json["id_token"] as? String, let rt = json["refresh_token"] as? String else {
            throw AuthError.shape("securetoken response missing id_token/refresh_token")
        }
        return (id, rt)
    }

    // MARK: - Password + TOTP MFA

    private func signInWithMfa(
        email: String, password: String, totpSecret: String
    ) async throws -> (String, String) {
        let pw = try await idpPost("v1/accounts:signInWithPassword", body: [
            "email": email,
            "password": password,
            "returnSecureToken": true,
        ])

        guard let pending = pw["mfaPendingCredential"] as? String else {
            throw AuthError.shape("no mfaPendingCredential (account not MFA-enrolled?)")
        }
        let mfaInfo = pw["mfaInfo"] as? [[String: Any]]
        guard let enrollmentId = mfaInfo?.first?["mfaEnrollmentId"] as? String else {
            throw AuthError.shape("no mfaInfo[0].mfaEnrollmentId in sign-in response")
        }

        let code = try await TOTP.freshCode(base32Secret: totpSecret)
        let final = try await idpPost("v2/accounts/mfaSignIn:finalize", body: [
            "mfaPendingCredential": pending,
            "mfaEnrollmentId": enrollmentId,
            "totpVerificationInfo": ["verificationCode": code],
        ])

        guard let id = final["idToken"] as? String, let rt = final["refreshToken"] as? String else {
            throw AuthError.shape("mfaSignIn:finalize response missing idToken/refreshToken")
        }
        return (id, rt)
    }

    // MARK: - Helpers

    private func idpPost(_ path: String, body: [String: Any]) async throws -> [String: Any] {
        guard let url = URL(string: "\(idpBase)/\(path)?key=\(apiKey)") else {
            throw AuthError.shape("bad IDP URL for \(path)")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        try ensureOK(response, data, path: path)
        return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
    }

    private func ensureOK(_ response: URLResponse, _ data: Data, path: String) throws {
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.shape("\(path): no HTTP response")
        }
        guard (200 ... 299).contains(http.statusCode) else {
            throw AuthError.http(path: path, status: http.statusCode, body: String(data: data, encoding: .utf8) ?? "")
        }
    }
}

File: windows/RecordHarness/FirebaseAuth.cs
using System.Text;
using System.Text.Json;

namespace RecordHarness;

/// <summary>
/// Minimal Firebase Identity Platform REST client for the harness's
/// self-contained sign-in, so the runner can authenticate the pinned test user
/// without Node or any prior e2e state.
///
/// Strategy: try the cached refresh token first (cheap, and it avoids the
/// rate-limited <c>mfaSignIn:finalize</c> quota); on ANY failure — 401, expired or
/// revoked refresh token, network error — fall back to the full TOTP MFA dance.
/// Either path returns a fresh, MFA-stamped ID token plus a NEW refresh token
/// (the old one is invalidated within seconds of exchange).
///
/// The C# port of the harness's Swift <c>FirebaseAuth</c>.
/// </summary>
public sealed class FirebaseAuth(string apiKey, HttpClient? http = null)
{
    private const string IdpBase = "https://identitytoolkit.googleapis.com";
    private const string SecureTokenBase = "https://securetoken.googleapis.com";

    private readonly HttpClient _http = http ?? new HttpClient();

    /// <param name="IdToken">Fresh, MFA-stamped ID token.</param>
    /// <param name="RefreshToken">New refresh token; the one passed in is now invalid.</param>
    /// <param name="Mode">Which path produced it — <c>refresh-exchange</c> or <c>totp-mfa</c>.</param>
    public sealed record MintResult(string IdToken, string RefreshToken, string Mode);

    /// <summary>
    /// Mints an ID token: refresh-token exchange when one is supplied and works,
    /// otherwise password + TOTP MFA.
    /// </summary>
    public async Task<MintResult> MintAsync(
        string? refreshToken,
        string? email,
        string? password,
        string? totpSecret,
        CancellationToken cancellationToken = default)
    {
        if (!string.IsNullOrEmpty(refreshToken))
        {
            try
            {
                var (id, rt) = await ExchangeRefreshTokenAsync(refreshToken, cancellationToken);
                return new MintResult(id, rt, "refresh-exchange");
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                // Expected whenever the cached token has aged out or been rotated;
                // the TOTP path below is the recovery, so this is a note, not a failure.
                Harness.Log($"refresh exchange failed, falling back to TOTP MFA: {ex.Message}");
            }
        }

        if (string.IsNullOrEmpty(email) || string.IsNullOrEmpty(password) || string.IsNullOrEmpty(totpSecret))
            throw new HarnessException("TOTP fallback requires FB_EMAIL, FB_PASSWORD, FB_TOTP_SECRET");

        var (idToken, newRefresh) = await SignInWithMfaAsync(email, password, totpSecret, cancellationToken);
        return new MintResult(idToken, newRefresh, "totp-mfa");
    }

    // --- Refresh-token exchange (securetoken) ---

    private async Task<(string IdToken, string RefreshToken)> ExchangeRefreshTokenAsync(
        string refreshToken, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(
            HttpMethod.Post, $"{SecureTokenBase}/v1/token?key={apiKey}")
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["grant_type"] = "refresh_token",
                ["refresh_token"] = refreshToken,
            }),
        };

        using var response = await _http.SendAsync(request, cancellationToken);
        var json = await ReadJsonAsync(response, "securetoken/v1/token", cancellationToken);

        // This endpoint returns snake_case despite the rest of IDP being camelCase.
        var id = json.GetPropertyOrNull("id_token")?.GetString();
        var rt = json.GetPropertyOrNull("refresh_token")?.GetString();
        if (string.IsNullOrEmpty(id) || string.IsNullOrEmpty(rt))
            throw new HarnessException("securetoken response missing id_token/refresh_token");

        return (id, rt);
    }

    // --- Password + TOTP MFA ---

    private async Task<(string IdToken, string RefreshToken)> SignInWithMfaAsync(
        string email, string password, string totpSecret, CancellationToken cancellationToken)
    {
        var signIn = await IdpPostAsync("v1/accounts:signInWithPassword", new
        {
            email,
            password,
            returnSecureToken = true,
        }, cancellationToken);

        var pending = signIn.GetPropertyOrNull("mfaPendingCredential")?.GetString();
        if (string.IsNullOrEmpty(pending))
            throw new HarnessException("no mfaPendingCredential (account not MFA-enrolled?)");

        var enrollmentId = signIn.GetPropertyOrNull("mfaInfo")
            ?.EnumerateArray().FirstOrDefault()
            .GetPropertyOrNull("mfaEnrollmentId")?.GetString();
        if (string.IsNullOrEmpty(enrollmentId))
            throw new HarnessException("no mfaInfo[0].mfaEnrollmentId in sign-in response");

        var code = await Totp.FreshCodeAsync(totpSecret, cancellationToken);
        var final = await IdpPostAsync("v2/accounts/mfaSignIn:finalize", new
        {
            mfaPendingCredential = pending,
            mfaEnrollmentId = enrollmentId,
            totpVerificationInfo = new { verificationCode = code },
        }, cancellationToken);

        var id = final.GetPropertyOrNull("idToken")?.GetString();
        var rt = final.GetPropertyOrNull("refreshToken")?.GetString();
        if (string.IsNullOrEmpty(id) || string.IsNullOrEmpty(rt))
            throw new HarnessException("mfaSignIn:finalize response missing idToken/refreshToken");

        return (id, rt);
    }

    // --- Helpers ---

    private async Task<JsonElement> IdpPostAsync(
        string path, object body, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{IdpBase}/{path}?key={apiKey}")
        {
            Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json"),
        };
        using var response = await _http.SendAsync(request, cancellationToken);
        return await ReadJsonAsync(response, path, cancellationToken);
    }

    private static async Task<JsonElement> ReadJsonAsync(
        HttpResponseMessage response, string path, CancellationToken cancellationToken)
    {
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        if (!response.IsSuccessStatusCode)
            throw new HarnessException($"{path} failed: {(int)response.StatusCode} {Truncate(body)}");

        try
        {
            // Clone so the element outlives the JsonDocument.
            using var doc = JsonDocument.Parse(body);
            return doc.RootElement.Clone();
        }
        catch (JsonException ex)
        {
            throw new HarnessException($"{path}: response was not JSON: {ex.Message}");
        }
    }

    /// <summary>
    /// Trims an error body before it reaches a log. Firebase error payloads are
    /// small, but they can echo the request — and the request carries a password.
    /// </summary>
    private static string Truncate(string body) =>
        body.Length <= 200 ? body : body[..200] + "…";
}
